### DISCUSSION TITLE
Built-in support for Property Based Tests (aka Fuzzing)

### DISCUSSION THREAD

**Discussion by dubzzz** [vitest-dev/vitest#2212] [Ideas]
Hey 👋,

_I'm Nicolas, core maintainer and main contributor to the project called  [fast-check](https://github.com/dubzzz/fast-check). The project does property based testing and reaches 8 million (~650k when I opened the ticket~) downloads a month at the moment._

**In a nutshell:**

Property Based Tests is another way to test code. Instead of writing _hard-coded_ tests (such as "startWith should return true for prefix='t' value='toto'"), we write tests following specs (such as "startWith should return true for any prefix and value=prefix+suffix").

I like to say that "it can detect the unexpected".

It can be pushed even further to detect race conditions, CVEs...

**Why it matters?**

Well, while classical tests are a must have for test frameworks, property based tests leverage them to next level by making tests able to find bugs earlier. Having such built-in in vitest would probably help the JavaScript community to embrace even more the technique (up to now all JavaScript property frameworks put together are gathering a bit less than 1M downloads per month with 650k for [fast-check](https://github.com/dubzzz/fast-check) alone).

Some people asked such support on my side: https://github.com/dubzzz/fast-check/issues/3054

**Suggestions of integration**

While the libraries can be used as standalone libraries next to any test framework, some binding making them better integrated with test runners can always help. For [fast-check](https://github.com/dubzzz/fast-check), I came up with an integration for Jest working as a standalone package providing an API rewiring Jest and fast-check for a better integration but as I'm starting one for Vitest I'm wondering if the best option would not be to just have it native.

My idea would be to come up with some kind of `it.prop` (see [related topic for the binding between Jest and fast-check](https://github.com/dubzzz/fast-check/issues/3303)).

Alternative way being: third-party package wiring vitest to fast-check such as [`@fast-check/jest`](https://github.com/dubzzz/fast-check/tree/main/packages/jest#readme) but for vitest or users keep doing it on their side.

👋 I opened the discussion mostly because I believe that providing such test technique directly within Vitest would help a lot the JavaScript community. If the idea of having such built-ins directly within Vitest is something that might be interested for Vitest, then I can probably give an hand and start an integration if I get some guidance and if an API is somehow accepted.

💡 Let me know if you need more context. I linked [fast-check](https://github.com/dubzzz/fast-check) so that you can see more in depth what it's about.

**Comment by jasikpark** [comment]
I've been using fast-check with vitest as much as possible! This could be great!

**Comment by dubzzz** [comment]
Following multiple requests to have a fuzzing approach natively in fast-check library, I started to dig a little bit more on the subject and thought back about this opened discussion. Actually, many people want **fuzzing** to be a thing in JavaScript and actually fast-check alone miss some parts of it, as vitest does. A good fuzzer (as the one of Go) requires a collaboration between the test runner and the fuzz system (fast-check is a fuzz system).

My point is: we should probably call the function `it.fuzz` instead of `it.prop`.

The normal test mode runs the test in a classical PBT manner. The fuzz mode could be inspired by what Go does with some kind of infinite runs or runs being capped by a time limit. Anyway, so far I implemented on a drafty version for the `it.fuzz` in test mode in [`@fast-check/vitest`](https://github.com/dubzzz/fast-check/tree/main/packages/vitest#readme), I'll start making my mind on a real fuzz mode. But from what I see: a closer collaboration with the test runner might be needed at some point to have a dedicated mode.

**Reply by sheremet-va** [reply]
> But from what I see: a closer collaboration with the test runner might be needed at some point to have a dedicated mode.

You can use a custom runner: https://vitest.dev/advanced/runner.html

**Reply by dubzzz** [reply]
Whaou, thanks for the link. I'll definitely check that one ❤️

**Reply by sheremet-va** [reply]
You can also create your own task function like `fuzz` that will be captured by `describe` - and it even has access to most of the options like `.todo` or `.skip`:

```ts
// fuzz.ts
import { getCurrentSuite, setFn, createChainable } from 'vitest/suite'

export const fuzz = createChainable(function fuzz(fc, name, fn) {
  const task = getCurrentSuite().custom.call(this, name)
  task.meta.fc = fc
  // this is needed only if you use default runner - you can track tasks however you prefer
  setFn(task, fn || (() => {}))
})

// test.ts
import { fuzz } from './fuzz.ts'

fuzz([fc.string(), fc.string(), fc.string()], 'should detect the substring', (a, b, c) => {
  return (a + b + c).includes(b);
})
```

**Reply by dubzzz** [reply]
I finally found a bit of time to start working on the suggestion you did. However I still struggle to make it work: I may miss something.

Definition of `@fast-check/vitest/fuzz`:

```ts
import { getCurrentSuite, setFn, createChainable } from 'vitest/suite';
import { assert, asyncProperty, readConfigureGlobal } from 'fast-check';
import type { Parameters as FcParameters } from 'fast-check';
import type { Prop, PromiseProp, ArbitraryTuple } from './internals/types.js';

function wrapProp<Ts extends [any] | any[]>(prop: Prop<Ts>): PromiseProp<Ts> {
  return (...args: Ts) => Promise.resolve(prop(...args));
}
export const fuzz = createChainable(['skip', 'only', 'todo'], function fuzz<
  Ts extends [any] | any[],
  TsParameters extends Ts = Ts,
>(this: any, name: string, arbitraries: ArbitraryTuple<Ts>, prop: Prop<Ts>, params?: FcParameters<TsParameters>): void {
  const currentSuite = getCurrentSuite();
  const task = currentSuite.custom.call(this, name);
  (task.meta as Record<string, unknown>) = { fuzz: true };
  setFn(task as any, async () => {
    const promiseProp = wrapProp(prop);
    const propertyInstance = (asyncProperty as any)(...(arbitraries as any), promiseProp);
    await assert(propertyInstance, params);
  });
});
```

My attempt to use it in `main-fuzz.spec.mjs`:

```ts
import { expect } from 'vitest';
import fc from 'fast-check';
import { fuzz } from '@fast-check/vitest/fuzz';

fuzz('should pass', [fc.constant(null)], (value) => {
  expect(value).toBe(null);
});
```

And the configuration for vitest in `vitest.config.mjs`:

```ts
import { defineConfig } from 'vite';

export default defineConfig({
  test: {
    include: ['test-bundle/main-fuzz.spec.mjs'],
  },
});
```

It seems that whatever I pass to `setFn`, it never gets executed. Even a function immediately throwing.

**Reply by sheremet-va** [reply]
Yes, looks like it doesn't correctly work at the moment. I found where the problem is, the fix should arrive shortly. But you can patch it here:

https://github.com/vitest-dev/vitest/blob/bf8393643ab3764ab9d727b6d064f2123832e82f/packages/runner/src/run.ts

```diff
-  if (c.type === 'test')
+  if (c.type === 'test' || c.type === 'custom')
```

And update your code to support some of the expected features:

```ts
const task = currentSuite.custom.call(this, name);
task.context = {};
```

**Reply by dubzzz** [reply]
Thanks a lot 😜

Other questions:

- Is there a way to access to the beforeEach function? As fuzzing means running many times, if there is a beforeEach defined I'll have to trigger it.
- A way to pass a custom timeout to the runner for this task? Actually the fuzzer will handle the stop by itself.

**Reply by sheremet-va** [reply]
> A way to pass a custom timeout to the runner for this task? Actually the fuzzer will handle the stop by itself.

You will need to wrap it into your own timeout function. You can pass it down however you want. In the next version you can just pass it down as an option.

You can ask this task in `beforeEach`:

```ts
beforeEach({ task }) => {
  if (task.meta.fuzz) {
    // ...
  }
})
```

Remember that meta can only contain serializable data (so, no functions). You can put functions on context (will be provided on a task in the next version). Basically, currently, the task is not compatible with test so some functionality doesn't work, but it will be fixed in the next version.

Updated way:

```ts
currentSuite.custom(name, {
   ...this,
   timeout: options?.timeout,
   meta: {
     fuzz: true
   },
   handler: async () => {
    const promiseProp = wrapProp(prop);
    const propertyInstance = (asyncProperty as any)(...(arbitraries as any), promiseProp);
    await assert(propertyInstance, params);
  },
});
```

**Reply by dubzzz** [reply]
For the `beforeEach` function, is it possible to trigger calls to it from the handler I declare on the task?

**Reply by sheremet-va** [reply]
Do you want to call it twice? I don't understand the usecase

**Reply by sheremet-va** [reply]
You can track the progress there: #4076

If you have any questions or suggestions feel free to leave it there.

**Reply by dubzzz** [reply]
> Do you want to call it twice? I don't understand the usecase

In the context of property based testing or fuzzing, the very same function will be executed multiple times (by default 100 times if everything goes well in fast-check). Basically the aim of these techniques is to generate inputs that may detect issues on the code under tests. Even if they are quite efficient to detect problems quickly they tend to require a certain number of runs (a hundred being very safe). As a consequence, the same function will be executed multiple times, so if beforeEach have been provided by the user it looks legit to run them.

Ideally, the handler code should be something like:

```ts
   handler: async () => {
    const promiseProp = wrapProp(prop);
    const propertyInstance = (asyncProperty as any)(...(arbitraries as any), promiseProp);
    await assert(
      propertyInstance.beforeEach(theBeforeEachDefinedForVitest).afterEach(theAfterEachDefinedForVitest),
      params
    );
  },
```

**Reply by sheremet-va** [reply]
Then you should provide 100 tests, this is how this system works. One `custom` (renamed to `task`) is one task, beforeEach/afterEach runs before and after each _task_. How you populate it is your implementation

**Reply by dubzzz** [reply]
I just tested https://github.com/vitest-dev/vitest/commit/e744d9e4110e9dbc54b8ada1c0a8982b90bdc51c by relying on v1.0.0-beta.0 of vitest it works perfectly!

---

**UPDATE:** I just have a typing issue with the output of `createChainable`. So far, I have not succeeded into typing it in a way the produced function (fuzz) can rely on generics. But I'll retry later. At the moment I have: `const fuzz: ChainableFunction<"concurrent" | "only" | "skip" | "todo" | "fails", [name: string, arbitraries: ArbitraryTuple<any[] | [any]>, prop: Prop<any[] | [any]>, params?: FcParameters<any[] | [any]> | undefined], void, {}>` while I'd need `T` to replace some of the `any`. But honestly I think most of the issue is on my side. Exposing `ChainableFunction` might help but I'm not even sure, I'll do some tests on my side.

My code today being:

```ts
export const fuzz = createChainable(['skip', 'only', 'todo', 'concurrent', 'fails'], function fuzz<
  Ts extends [any] | any[],
  TsParameters extends Ts = Ts,
>(this: any, name: string, arbitraries: ArbitraryTuple<Ts>, prop: Prop<Ts>, params?: FcParameters<TsParameters>): void {...}
```

**Reply by sheremet-va** [reply]
> I just have a typing issue with the output of createChainable

I would recommend using the new "createTestCollector" utility instead (it implements more APIs like skipIf, but also extend and each which might not be desirable in your cade). Although you will probably have similar issues with types there 🤔

**Reply by dubzzz** [reply]
I have taken the time to rework on the vitest integration since our last interaction. I was fighting with making v4 of the package. By the way I just updated the description of the issue as since the opening of it fast-check passed from 700k to 8 million monthly downloads.

**Reply by sheremet-va** [reply]
Just a note - it's probably better to provide `it.fuzz` instead of a `fuzz`. This will make static analysis easier 😄

**Reply by dubzzz** [reply]
I have it.fuzz like syntax already in place for now. Perfect 👌

**Comment by bertschneider** [comment]
Even though it's a somewhat different approach, in the project [Jazzer.js](https://github.com/CodeIntelligenceTesting/jazzer.js) we combined [libFuzzer with Jest](https://github.com/CodeIntelligenceTesting/jazzer.js#using-test-framework-integration) and thought about providing a similar [integration with vitest](https://github.com/CodeIntelligenceTesting/jazzer.js/issues/343).

Perhaps we could collaborate on a vitest integrations to get the best of both approaches.

Using `it.fuzz` feels quite natural, so I would encourage using that syntax.

**Comment by dubzzz** [comment]
Time passed since my last update on this discussion. Since the last message I thought about other ideas.

I recently drafted and released a new proposal for integrating property based testing capacities as a first class citizen within a test runner like Vitest. Here is the writing I did to motivate that idea: https://fast-check.dev/blog/2025/03/28/beyond-flaky-tests-bringing-controlled-randomness-to-vitest/

**Comment by ahrjarrett** [comment]
Would a **huge** step for the industry as a whole to see vitest support property testing natively.

If history is any indication, and _especially_ with the explosion of "let AI write your unit tests for you" ads I see everywhere these days, I'd say the JavaScript ecosystem is overdue for property testing to become more mainstream.

And if vitest is going to support property testing, I would also urge the library authors to resist the urge to implement their own version. `fast-check` is one of those rare libraries that has managed to find the sweet spot between theory, and practicality.

**Comment by waynevanson** [comment]
An idea for the pool.

A tighter integration between Vitest and property based testing means Vitest could manage "failure persistence" as described in Rust's `proptest`.

https://proptest-rs.github.io/proptest/proptest/failure-persistence.html

I see the easiest way to implement this today could be a Vitest runner that reads this persistence file and starts and injects these seeds first.

**Reply by ahrjarrett** [reply]
I was wondering about something similar myself at one point. My thought was that it would be convenient if a tool could write the failing case to `examples`, similar to how Vitest handles inline snapshots.

Writing to disc is a better idea – otherwise the examples would get bigger over time, and make the test less readable.
